packages/language-server/src/plugins/typescript/features/DocumentSymbolsProvider.ts

```diff
@@ -0,0 +1,100 @@
+import { NavigationTree } from 'typescript';
+import { SymbolInformation, Range, SymbolKind } from 'vscode-languageserver-types';
+import { AstroDocument, mapSymbolInformationToOriginal } from '../../../core/documents';
+import { DocumentSymbolsProvider } from '../../interfaces';
+import { LanguageServiceManager } from '../LanguageServiceManager';
+import { SnapshotFragment } from '../snapshots/DocumentSnapshot';
+import { symbolKindFromString } from '../utils';
+
+export class DocumentSymbolsProviderImpl implements DocumentSymbolsProvider {
+	constructor(private languageServiceManager: LanguageServiceManager) {}
+
+	async getDocumentSymbols(document: AstroDocument): Promise<SymbolInformation[]> {
+		const { lang, tsDoc } = await this.languageServiceManager.getLSAndTSDoc(document);
+		const fragment = await tsDoc.createFragment();
+
+		const navTree = lang.getNavigationTree(tsDoc.filePath);
+
+		const symbols: SymbolInformation[] = [];
+		this.collectSymbols(navTree, fragment, undefined, (symbol) => symbols.push(symbol));
+
+		const result: SymbolInformation[] = [];
+
+		// Add a "frontmatter" namespace for the frontmatter if we have a closed one
+		if (document.astroMeta.frontmatter.state === 'closed') {
+			result.push(
+				SymbolInformation.create(
+					'Frontmatter',
+					SymbolKind.Namespace,
```

Comment:
Used the Namespace SymbolKind here for two reasons:
- It made the most sense to me, it also look the prettiest with the {} icon
- It seems to be the one most extension uses for this kind of symbols (ex: Volar does it for the different parts of a Vue file). Svelte uses field but I think it's only because it's the default one for HTML elements (and Svelte like to consider the different parts of its files to be "HTML elements")